extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java

```diff
@@ -99,13 +118,19 @@ public void killAll() throws IOException
     fs.delete(storageDirectory, true);
   }
 
-  private boolean safeNonRecursiveDelete(FileSystem fs, Path path)
+  private void deleteEmptyParents(FileSystem fs, Path path)
   {
+    if(fs.makeQualified(path).equals(fs.makeQualified(storageDirectory))) {
```

Comment:
This check is weak since users can change storageDirectory at any time, and that will affect new segments, but old segments will remain in their old locations and still be killable there. So it's better for single-segment killing logic to avoid depending on storageDirectory in any way.

Maybe use the "version", "interval", and "dataSource" fields of the segment to detect when to stop walking back up the parents?